i18n/create_json.js

```diff
@@ -79,43 +80,77 @@ const parseStringFile = function (jsonstream) {
+      } else {
+        result = result.replace(`$\{${var_name}}`, defdata[var_name]);
+      }
     }
   });
   return result;
 }
 
 /**
- * main stuff
+ * Read catroid string file and parse it into a js object
+ * @param {*} file 
+ * @param {*} enc 
+ * @return Object
+ */
+const readStringFile = async (file, enc = 'utf-8') => {
+  const payload = fs.readFileSync(file, { encoding: enc });
+  return parseStringFile(await xml2json.parseStringPromise(payload));
+}
+
+/**
+ * Generate language mapping
+ * @param {*} values 
+ * @return Object
+ */
+const generateLangValues = (values, defvalues) => {
+  const result = {};
+  Object.keys(MAPPING).filter(key => !key.startsWith(MAPPING_COMMENT)).forEach(rule => {
+    const value = substituteVariableData(MAPPING[rule], values, defvalues);
+    result[rule] = value.split('"').join('');
+  });
+
+  return result;
+}
+
+/**
+ * Create language file based on MAPPING and values
+ *  Dump result as json file into JSON_DIR
+ * @param {*} lang 
+ * @param {*} values 
  */
-const mapping = JSON.parse(
-  fs.readFileSync(MAPPING_FILE, { encoding: 'utf-8' })
-);
-const languages = fs.readdirSync(STRINGS_DIR, { encoding: 'utf-8' });
-
-languages.forEach(language => {
-  const lang_file_path = path.join(STRINGS_DIR, language, STRINGS_FILE);
-  const lang_file = fs.readFileSync(lang_file_path, { encoding: 'utf-8' });
-  xml2json.parseStringPromise(lang_file).then(data => {
-    const lang_values = parseStringFile(data);
-    const lang_name = prepareStringFolderName(language);
-    const json_file = path.join(JSON_DIR, lang_name + '.json');
-
-    const result = {};
-
-    Object.keys(mapping).filter(key => !key.startsWith(MAPPING_COMMENT)).forEach(rule => {
-      const value = substituteVariableData(mapping[rule], lang_values);
-      result[rule] = value.split('"').join('');
-    });
-    result["DROPDOWN_NAME"] = locales[lang_name] ? locales[lang_name] : lang_name;
-
-    fs.writeFileSync(json_file, JSON.stringify(result), { encoding: 'utf-8' });
+const dumpStringFile = (lang, values) => {
+  const langname = prepareStringFolderName(lang);
+  const jsonfile = path.join(JSON_DIR, langname + '.json');
+  values["DROPDOWN_NAME"] = locales[langname] ? locales[langname] : langname;
+  fs.writeFileSync(jsonfile, JSON.stringify(values), { encoding: 'utf-8' });
+}
+
+
+(async () => {
+  if (!fs.existsSync(JSON_DIR)) fs.mkdirSync(JSON_DIR);
+
+  const def_string = await readStringFile(DEF_LANG_FILE);
+  const def_values = generateLangValues(def_string, def_string);
+  dumpStringFile(DEF_LANG, def_values);
+
+  const languages = fs.readdirSync(STRINGS_DIR, { encoding: 'utf-8' });
+  languages.filter(lang => (lang != DEF_LANG)).forEach(async langname => {
+    const lang_file = path.join(STRINGS_DIR, langname, STRINGS_FILE);
+    const lang_string = await readStringFile(lang_file);
+    const lang_values = generateLangValues(lang_string, def_string);
+    dumpStringFile(langname, lang_values);
   });
```

Comment:
refactored existing code to use async and await. Makes the code more cleaner. In addition it is quit easy to see where we load the default language and pass it to each further lang.